Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs

```diff
@@ -319,70 +325,85 @@ private static async Task<RequestModel> DecryptEncryptedRequestFromShiftsAsync(b
         {
             List<ShiftsIntegResponse> responseModelList = new List<ShiftsIntegResponse>();
             var requestBody = jsonModel.Requests.First(x => x.Url.Contains("/openshiftrequests/", StringComparison.InvariantCulture)).Body;
-            var requestState = requestBody != null ? requestBody["state"].Value<string>() : null;
 
             if (requestBody != null)
             {
+                var requestState = requestBody?["state"].Value<string>();
+
                 switch (requestState)
                 {
                     // The Open shift request is submitted in Shifts and is pending with manager for approval.
                     case ApiConstants.ShiftsPending:
                         {
-                            responseModelList = await this.ProcessOutboundOpenShiftRequestAsync(jsonModel, updateProps).ConfigureAwait(false);
+                            var openShiftRequest = JsonConvert.DeserializeObject<OpenShiftRequestIS>(requestBody.ToString());
+                            responseModelList = await this.ProcessOutboundOpenShiftRequestAsync(openShiftRequest, updateProps, teamsId).ConfigureAwait(false);
                         }
 
                         break;
 
                     // The Open shift request is approved by manager.
                     case ApiConstants.ShiftsApproved:
                         {
-                            responseModelList = await this.ProcessOpenShiftRequestApprovalAsync(jsonModel, updateProps).ConfigureAwait(false);
+                            // Passthrough header not present for outbound call i.e the request is coming from either Shifts or some unknown source.
+                            if (string.IsNullOrEmpty(passThroughHeaderValue))
+                            {
+                                this.telemetryClient.TrackTrace(Resource.PassthroughHeaderNotPresent);
+                                responseModelList = this.GenerateResponseToPreventAction(jsonModel, Resource.PreventManageActionFromShifts);
+                            }
+
+                            // Manager has approved open shift request in Kronos.
+                            else if (passThroughHeaderValue.Equals(workforceIntegrationId, StringComparison.Ordinal))
+                            {
+                                this.telemetryClient.TrackTrace(Resource.PassthroughHeaderWFIIDMatched);
+                                responseModelList = await this.ProcessOpenShiftRequestApprovalAsync(jsonModel, updateProps).ConfigureAwait(false);
+                            }
+
+                            // Passthrough header is present but the request is coming from some different workforce management.
+                            else
+                            {
+                                this.telemetryClient.TrackTrace(Resource.PassthroughHeaderWFIIDMismatch);
```

Comment:
This too you can move up to the beginning of controller.

Reply:
@aosolis , We wanted to implement passthrough header for OpenShiftRequest and SwapShiftRequest only. As in other cases (Shifts and Open Shifts) we have to simply send the response as it can be a genuine request from Shifts's other entity and also when a request is being made from Shifts UI (for open shift request and swap shift request), we don't get any passthrough header as part of outbound call.

Reply:
Why would you implement the passthrough header for certain entities only? The header is used to signal that this is a request that your integration initiated, so there should be no further need to synchronize it back to Kronos, because the notification is either
1. From a change that is being synced from Kronos->Shifts, or
2. From a change that you made in response to an earlier notification from Shifts that your integration is in the process of handling (and *that* handler should take care of ensuring Shifts<->Kronos consistency)